src/librustc_passes/check_const.rs

```diff
@@ -0,0 +1,136 @@
+//! This pass checks the HIR bodies of const items (e.g., `const`, `static`, `const fn`) for
+//! structured control flow (e.g. `if`, `while`), which is forbidden in a const context.
+//!
+//! By the time the MIR const-checker runs, these high-level constructs have been lowered to
+//! control-flow primitives (e.g., `Goto`, `SwitchInt`), making it tough to properly attribute
+//! errors. We still look for those primitives in the MIR const-checker to ensure nothing slips
+//! through, but errors for structured control flow in a `const` should be emitted here.
+
+use rustc::hir::def_id::DefId;
+use rustc::hir::intravisit::{Visitor, NestedVisitorMap};
+use rustc::hir::map::Map;
+use rustc::hir;
+use rustc::session::Session;
+use rustc::ty::TyCtxt;
+use rustc::ty::query::Providers;
+use syntax::span_err;
+use syntax_pos::Span;
+
+fn check_mod_const_bodies(tcx: TyCtxt<'_>, module_def_id: DefId) {
+    if tcx.sess.opts.debugging_opts.unleash_the_miri_inside_of_you {
+        return;
+    }
+
+    let mut vis = CheckConstVisitor::new(tcx);
+    tcx.hir().visit_item_likes_in_module(module_def_id, &mut vis.as_deep_visitor());
+}
+
+pub(crate) fn provide(providers: &mut Providers<'_>) {
```

Comment:
```suggestion
crate fn provide(providers: &mut Providers<'_>) {
```

Reply:
This would require adding a feature flag to this crate. I'll wait for an automated refactoring to do this.

Reply:
Fair. :) just as a note for the future tho: it's entirely fine to add that feature gate and take advantage of nightly features.